src/middleware/http-exceptions.ts

```diff
@@ -0,0 +1,105 @@
+export type HttpExceptionMetadata = {
+  type: string
+  message: string
+  data?: Record<string, unknown>
+} & Record<string, unknown>
+
+export interface ThrowingOptions {
+  json?: boolean
+}
+
+/**
+ * Throw HttpExceptions inside API endpoints to generate nice error messages
+ *
+ * @example
+ * ```
+ * if (bad_soups.includes(soup_param)) {
+ *   throw new HttpException(400, {
+ *      type: "cant_make_soup",
+ *      message: "Soup was too difficult, please specify a different soup"
+ *    })
+ * }
+ * ```
+ *
+ **/
+export class HttpException extends Error {
+  options: ThrowingOptions
+
+  constructor(
+    public status: number,
+    public metadata: HttpExceptionMetadata,
+    options?: ThrowingOptions
+  ) {
+    super(metadata.message)
+
+    if (Error.captureStackTrace) {
+      Error.captureStackTrace(this, this.constructor)
+    }
+
+    this.options = options ?? { json: true }
+  }
+
+  toString() {
+    return `HttpException: ${this.status}, ${this.metadata.message} (${this.metadata.type})`
+  }
+}
+
+/**
+ * Throw BadRequestException inside API endpoints that were provided incorrect
+ * parameters or body
+ *
+ * @example
+ * ```
+ * if (bad_soups.includes(soup_param)) {
+ *   throw new BadRequestException({
+ *     type: "cant_make_soup",
+ *     message: "Soup was too difficult, please specify a different soup",
+ *   })
+ * }
+ * ```
+ *
+ **/
+export class BadRequestException extends HttpException {
+  constructor(
+    public metadata: HttpExceptionMetadata,
+    options?: ThrowingOptions
+  ) {
+    super(400, metadata, options)
+  }
+}
+
+export class UnauthorizedException extends HttpException {
+  constructor(
+    public metadata: HttpExceptionMetadata,
+    options?: ThrowingOptions
+  ) {
+    super(401, metadata, options)
+  }
+}
+
+export class NotFoundException extends HttpException {
+  constructor(
+    public metadata: HttpExceptionMetadata,
+    options?: ThrowingOptions
+  ) {
+    super(404, metadata, options)
+  }
+}
+
+export class MethodNotAllowedException extends HttpException {
+  constructor(
+    public metadata: HttpExceptionMetadata,
+    options?: ThrowingOptions
+  ) {
+    super(405, metadata, options)
+  }
+}
+
+export class InternalServerErrorException extends HttpException {
+  constructor(
+    public metadata: HttpExceptionMetadata,
+    options?: ThrowingOptions
+  ) {
+    super(500, metadata, options)
+  }
+}
```

Comment:
imo these should probably be at the application level rather than the framework, everyone may not like the default http error formatting

and internally for fakes we often need different error response shapes than the default

maybe instead of handling these errors by default, export a middleware that users can add to explicitly handle these?

Reply:
What do we throw from our validation middleware? Is a custom validation error alright?

Reply:
yeah good point, was also thinking about that as I scrolled down haha

maybe a new error class, `EdgeSpecDefaultMiddlewareError`, that internal middleware errors extend from? then when they throw we can return a 500 and as long as something is logged to the console explaining what it is and how to return better errors by adding more middleware I think we're good

if we adopted a HTTP error `interface` too (would highly recommend) then errors extending  `EdgeSpecDefaultMiddlewareError` could be automatically handled if the user chooses to import and use our default error middleware

Reply:
yeah I was thinking something like this

Reply:
> Current plan: add a top-level middleware, that basically handles all un-handled errors, returns 500 with empty body, and if they are EdgeSpecDefaultMiddlewareError print to console saying basically "EdgeSpecDefaultMiddlewareError is unhandled, consider using the default middleware error handler"

sounds perfect to me 👍

Reply:
did the above